staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go

```diff
@@ -660,6 +697,170 @@ See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts`
 	return nil
 }
 
+// Saves the last-applied-configuration annotation in a separate SSA field manager
+// to prevent it from being dropped by users who have transitioned to SSA.
+//
+// If this operation is not performed, then the last-applied-configuration annotation
+// would be removed from the object upon the first SSA usage.
+//
+// If the annotation is not already present, or if it is already managed by the
+// separate SSA fieldmanager, this is a no-op.
+func (o *ApplyOptions) saveLastApplyAnnotationIfNecessary(
+	helper *resource.Helper,
+	info *resource.Info,
+) error {
+	// Send an apply patch with the last-applied-annotation
+	// so that it is not orphaned by SSA in the following patch:
+	accessor, err := meta.Accessor(info.Object)
+	if err != nil {
+		return err
+	}
+
+	// Get the current annotations from the object.
+	annots := accessor.GetAnnotations()
+	if annots == nil {
+		annots = map[string]string{}
+	}
+
+	fieldManager := fieldManagerLastAppliedAnnotation
+	originalAnnotation, hasAnnotation := annots[corev1.LastAppliedConfigAnnotation]
+
+	// If the annotation does not already exist, we do not do anything
+	if !hasAnnotation {
```

Comment:
For this one, I'm wondering if we should also skip if the _current_ field manager name isn't `kubectl`. The current behaviour always preserves the annotation because it continues to be owned by the CSA manager, BUT the downgrade doesn't actually work, because the annotation is not being kept in sync. So it would be a behaviour change... but arguably the retention when it is not kept correct is a bug. Wdyt?

Reply:
To be clear, this is a suggestion to drop the annotation if during the migration the user is not using the default `kubectl` field manager? I am on board but want to be clear that this entails the annotation gets dropped if at the time of the first migration, the field manager is changed.

Reply:
> want to be clear that this entails the annotation gets dropped if the field manager is not kubectl at the time of the first migration.

Yes, that's what I'd expect. Can you think of any reason it would be valuable to retain in that case? The docs explicitly say that custom field managers opt out of the downgrade capability. Imagining the following sequence:
1. `kubectl apply -f mycm.yaml`
2. `kubectl apply -f mycm.yaml --server-side --field-manager=custom`
3. `kubectl apply -f mycm.yaml`

If step 2 made any changes, step 3 is going to use broken logic whether or not we keep the annotation, because the data it is using is either empty (behaves like a fresh apply) or stale (behaviour undefined).  I guess if they instead do:

1. `kubectl apply -f mycm.yaml`
2. `kubectl apply -f mycm.yaml --server-side --field-manager=custom`
3. `kubectl apply -f mycm.yaml --server-side --force-conflicts`
4. `kubectl apply -f mycm.yaml`

Then step 3 will repair the annotation if it is still present? I guess that could be considered useful, but I don't like the tradeoff of essentially undefined behaviour (instead of "this behaves like a fresh apply") in the first workflow, which I feel would be more common.

(fwiw I don't think my other argument re: users relying on it for miscellaneous reasons applies here, because for one thing they are actively opting in, and for another, invalid data is at best useless and at worst harmful.)

Reply:
>  Can you think of any reason it would be valuable to retain in that case?

One consideration is that in the current form if an applier made use of the the `kubectl` field manager later, the annotation would be updated once again. 

> Then step 3 will repair the annotation if it is still present? 

Step 2 would have caused the annotation to be dropped since it would have no owner. It would be gone by #3.

Reply:
> One consideration is that in the current form if an applier made use of the the kubectl field manager later, the annotation would be updated once again.

If they really want that, would it work for them to add the annotation in with a blank value themselves? Experimentally, it seems to (create an object with SSA and then add `kubectl.kubernetes.io/last-applied-configuration: "{}"` to it--suddenly it will be populated!)

> Step 2 would have caused the annotation to be dropped since it would have no owner. It would be gone by no. 3.

If we don't opt it out of our migration process here, then the special field manager would be keeping it around in 2 and 3, no?

Reply:
Considering the above, agreed it is best to not migrate the annotation if custom field manager is detected. Since its value would always be wrong anyway.

Reply:
> If we don't opt it out of our migration process here, then the special field manager would be keeping it around in 2 and 3, no?

when a custom field manager is used:

if you make the change you propose and DO opt it out of the migration process: step 2 would drop annotation
if we do nothing, and the annotation is applied into a new separate field manager on behalf of the user: both step 2 and step 3 retain the annotation

Reply:
Sounds good. Can we also add a test (or maybe just augment the existing non-default manager test) to show we're removing the unmaintained data on purpose?

Reply:
logic has been added and test augmented